src/msg/async/Event.cc

```diff
@@ -452,12 +452,16 @@ int EventCenter::process_events(unsigned timeout_microseconds,  ceph::timespan *
 
 void EventCenter::dispatch_event_external(EventCallbackRef e)
 {
-  external_lock.lock();
-  external_events.push_back(e);
-  bool wake = !external_num_events.load();
-  uint64_t num = ++external_num_events;
-  external_lock.unlock();
-  if (!in_thread() && wake)
+  uint64_t num = 0;
+  {
+    std::lock_guard lock{external_lock};
+    if (external_num_events > 0 && *external_events.rbegin() == e) {
+      return;
+    }
+    external_events.push_back(e);
+    num = ++external_num_events;
+  }
+  if (!in_thread() && 1 == num)
```

Comment:
i'd suggest put
```c++
if (num == 1 && !in_thread()) {
```
for better performance and to be more consistent. yoda conditions are not forbidden but i failed to find its occurrences in https://google.github.io/styleguide/cppguide.html#Conditionals.

Reply:
ok, that's better, thx for your remind.